tests/unit/test_compute_linear_bias.py

```diff
@@ -0,0 +1,45 @@
+import torch
+
+from transformer_lens.utils import expand_alibi_on_query_dim
+
+
+def test_expand_alibi_on_query_dim():
+    # Define the tensor x
+    x = torch.tensor([[[[1, 2, 3, 4, 5, 6, 7]]]])  # shape: (1, 1, 1, 7)
+    query_pos = 7
+    result = expand_alibi_on_query_dim(x, query_pos)
+
+    for i in range(query_pos):
+        # The first i values should be decreasing from 7 to 7-i+1
+        expected_first_values = torch.arange(7 - i + 1, 8, device=x.device)
+
+        # The remaining values should be 7
+        expected_remaining_values = torch.tensor(
+            [7] * (query_pos - len(expected_first_values)), device=x.device
+        )
+
+        # Concatenate to form the expected row
+        expected_row = torch.cat((expected_first_values, expected_remaining_values))
+
+        # Check if the result matches the expected row
+        if not torch.equal(result[0, 0, i], expected_row):
+            return False
```

Comment:
I've put some suggestions for how to simplify the underlying functions in the components file.

If you agree with that approach, the functions are now small enough to make simple unit tests that follow a different approach to the main code (i.e. the tests can use loops rather than vectorised tensor operations

An example for this could be testing create_alibi_slope:

```
n_ctx = 100

# Calculate the expected slope, with a non-vectorised approach
slope_expected = torch.zeros((n_ctx, n_ctx))
for row in range(n_ctx):
    for col in range(n_ctx):
        # Create the slope value
        slope_expected[row, col] = float(min(col - row, 0))

# Check against the method's vectorised version
result = Attention.create_alibi_slope(n_ctx)
assert torch.allclose(slope_expected, result)
```